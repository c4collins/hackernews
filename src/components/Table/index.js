// Core Modules
import React, { Component } from 'react';

// Third-Party Modules
import classNames from 'classnames';
import PropTypes from 'prop-types';

// Internal Modules
import Button from '../Button';

// Misc Files
import './index.css';

// Constants
import {
  SORTS,
  FontAwesome,
} from '../../constants';

class Table extends Component { 

    constructor(props){
        super(props);

        this.state = {
            sortKey: 'NONE',
            isSortReverse: false,
        };

        this.onSort = this.onSort.bind(this);
    }

    onSort(sortKey){
        const isSortReverse = this.state.sortKey === sortKey && !this.state.isSortReverse;
        this.setState({ sortKey, isSortReverse });
    }

    render() {
        const {
            list, 
            onDismiss,
        } = this.props;
        const {
            sortKey,
            isSortReverse
        } = this.state;

        const sortedList = SORTS[sortKey](list);
        const reverseSortedList = isSortReverse
            ? sortedList.reverse()
            : sortedList;

        return (
            <div className="table">
            <div className="table-header">
                <span style={{ width: '40%' }}>
                <Sort
                    sortKey={'TITLE'}
                    onSort={this.onSort}
                    activeSortKey={sortKey}
                    isSortReverse={isSortReverse}
                > Title
                </Sort>
                </span>
                <span style={{ width: '30%' }}>
                <Sort
                    sortKey={'AUTHOR'}
                    onSort={this.onSort}
                    activeSortKey={sortKey}
                    isSortReverse={isSortReverse}
                > Author
                </Sort>
                </span>
                <span style={{ width: '10%' }}>
                <Sort
                    sortKey={'COMMENTS'}
                    onSort={this.onSort}
                    activeSortKey={sortKey}
                    isSortReverse={isSortReverse}
                > Comments
                </Sort>
                </span>
                <span style={{ width: '10%' }}>
                <Sort
                    sortKey={'POINTS'}
                    onSort={this.onSort}
                    activeSortKey={sortKey}
                    isSortReverse={isSortReverse}
                > Points
                </Sort>
                </span>
                <span style={{ width: '10%' }}></span>
            </div>
            { reverseSortedList.map((item) =>
                <div key={item.objectID} className="table-row">
                <span style={{ width: '40%' }}>
                    <a href={item.url}>{item.title}</a>
                </span>
                <span style={{ width: '30%' }}>{item.author}</span>
                <span style={{ width: '10%' }}>{item.num_comments}</span>
                <span style={{ width: '10%' }}>{item.points}</span>
                <span style={{ width: '10%' }}>
                    <Button 
                    onClick={() => onDismiss(item.objectID)}
                    className="button-inline"
                    >
                    Dismiss
                    </Button>                
                </span>
                </div>
            )}
            </div>
        )
    }
}

Table.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      objectID: PropTypes.string.isRequired,
      author: PropTypes.string,
      url: PropTypes.string,
      num_comments: PropTypes.number,
      points: PropTypes.number,
    })
  ).isRequired,
  onDismiss: PropTypes.func.isRequired,
  sortKey: PropTypes.string.isRequired,
  onSort: PropTypes.func,
  isSortReverse: PropTypes.bool
};

Table.defaultProps = {
    sortKey: 'NONE'
}


const Sort = ({
  sortKey, 
  activeSortKey, 
  onSort, 
  isSortReverse,
  children 
}) => {
  const isSortActive = sortKey === activeSortKey;
  const sortClass = classNames(
    'button-inline',
    { 'button-active': isSortActive}
  );
  const iconClass = classNames(
    {'arrow-up': isSortReverse && isSortActive},
    {'arrow-down': !isSortReverse && isSortActive}
  )

  return (
    <Button 
      onClick={()=> onSort(sortKey)}
      className={sortClass}
    >
      <FontAwesome
        name={iconClass}
        style={{ 
          textShadow: '0 1px 0 rgba(0, 0, 0, 0.1)',
          color: 'black',
        }}
      />
      {children}
    </Button>
  );
}

Sort.propTypes = {
  sortKey: PropTypes.string.isRequired,
  activeSortKey: PropTypes.string,
  onSort: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
}


export default Table;
export { Sort };