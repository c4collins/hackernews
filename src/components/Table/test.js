import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Table, { Sort } from './index';

describe('Sort', () => {

  it('renders', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <Sort 
        onSort={() => null} 
        sortKey='NONE'
      >
        More
      </Sort>, div);
  });

  test('snapshots', () => {
    const component = renderer.create(
      <Sort 
        onSort={() => null} 
        sortKey='NONE'
      >
        More
      </Sort>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  
});

describe('Table', () => {

  const props = {
    list: [
      { title: '1', author: '1', num_comments: 1, points: 2, objectID: 'x'},
      { title: '2', author: '2', num_comments: 1, points: 2, objectID: 'y'},
    ],
    onDismiss: () => null,
    sortKey: 'NONE',
    onSort: () => null,
    isSortReverse: false
  }

  it('renders', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Table {...props} />, div);
  });

  it('shows two items in a list', () => {
    const element = shallow(
      <Table {...props} />
    );

    expect(element.find('.table-row').length).toBe(2);
  });

  test('snapshots', () => {
    const component = renderer.create(
      <Table {...props} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  
});