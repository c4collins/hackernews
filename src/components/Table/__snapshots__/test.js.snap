// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort snapshots 1`] = `
<button
  className="button-inline"
  onClick={[Function]}
  type="button"
>
  <span
    aria-hidden={true}
    className="fa fa-"
    style={
      Object {
        "color": "black",
        "textShadow": "0 1px 0 rgba(0, 0, 0, 0.1)",
      }
    }
  />
  More
</button>
`;

exports[`Table snapshots 1`] = `
<div
  className="table"
>
  <div
    className="table-header"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        <span
          aria-hidden={true}
          className="fa fa-"
          style={
            Object {
              "color": "black",
              "textShadow": "0 1px 0 rgba(0, 0, 0, 0.1)",
            }
          }
        />
         Title
      </button>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        <span
          aria-hidden={true}
          className="fa fa-"
          style={
            Object {
              "color": "black",
              "textShadow": "0 1px 0 rgba(0, 0, 0, 0.1)",
            }
          }
        />
         Author
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        <span
          aria-hidden={true}
          className="fa fa-"
          style={
            Object {
              "color": "black",
              "textShadow": "0 1px 0 rgba(0, 0, 0, 0.1)",
            }
          }
        />
         Comments
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        <span
          aria-hidden={true}
          className="fa fa-"
          style={
            Object {
              "color": "black",
              "textShadow": "0 1px 0 rgba(0, 0, 0, 0.1)",
            }
          }
        />
         Points
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    />
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href={undefined}
      >
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href={undefined}
      >
        2
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
